README.md
[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/XGDElements/xgd-table-of-contents)

# \<xgd-table-of-contents\>

Provides a Polymer 2 table of contents for use in a document library or other such use.

### Styling

The following custom properties and mixins are available for styling.

| Custom property | Description | Default |
| --- | --- | --- |
| `--toc-font-size` | Font size for the menus | 1em |
| `--app-primary-color` | Primary Color of your App | #000 |
| `--app-dark-primary-color` | Dark Version Primary Color of your App | #000 |
| `--app-lite-primary-color` | Light Version Primary Color of your App. | #fff |


Example:
```html
<xgd-table-of-contents id="toc"
    linkicon="icons:chevron-right" 
    opened="icons:expand-more" 
    closed="icons:expand-less">
</xgd-table-of-contents>

<script>
    var data = {
        "Title": "Table Of Contents",
        "Link": "",
        "Chapters": [
        {
            "Title": "Chapter A",
            "Link": "",
            "Chapters": [
                {
                    "Title": "Section 1",
                    "Link": "http://d.com"
                },
                {
                    "Title": "Section 2",
                    "Link": "http://z.com"
                },
                {
                    "Title": "Section 3",
                    "Link": "http://ddddd.com"
                }
            ]
        },
        {
            "Title": "Chapter 2",
            "Link": "",
            "Chapters": [
                {
                    "Title": "Section 1",
                    "Link": "http://aaa.com"
                },
                {
                    "Title": "Section 2",
                    "Link": "http://sss.com"
                },
                {
                    "Title": "Section 3",
                    "Link": "http://ddd.com"
                }
            ]
        }
      ]
    }
    this.$.toc.LoadTableOfContents(data);
</script>
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [v3.1.4](https://github.com/drumr8403/xgd-table-of-contents/compare/v3.1.4...v3.1.4) - 2018-08-30

## [v3.1.3](https://github.com/drumr8403/xgd-table-of-contents/compare/v3.1.2...v3.1.3) - 2018-08-29

### Commits

- Fixed selecting first link in chapter when selecting next button [`6f6b062`](https://github.com/drumr8403/xgd-table-of-contents/commit/6f6b062c0102bc5a3609b489a7daf2247e0c2d0e)

## [v3.1.2](https://github.com/drumr8403/xgd-table-of-contents/compare/v3.1.1...v3.1.2) - 2018-08-29

### Commits

- Fixed performance issue with TOC.  Credit ==> Grant Gryczan [`feff576`](https://github.com/drumr8403/xgd-table-of-contents/commit/feff57672c746ce3a39231798e7097477ae08716)

## [v3.1.1](https://github.com/drumr8403/xgd-table-of-contents/compare/v3.1.0...v3.1.1) - 2018-08-10

### Commits

- test for empty toc on load [`da62d7d`](https://github.com/drumr8403/xgd-table-of-contents/commit/da62d7dc0eb4a2b34dfbfdb8265fea1715a16b5e)

## [v3.1.0](https://github.com/drumr8403/xgd-table-of-contents/compare/v3.0.0...v3.1.0) - 2018-08-06

### Commits

- fixed error with polymer 2 conversion [`ea0a0dd`](https://github.com/drumr8403/xgd-table-of-contents/commit/ea0a0ddab8276342b379595ec1287bb0863ac0bb)

## [v3.0.0](https://github.com/drumr8403/xgd-table-of-contents/compare/1.0.2...v3.0.0) - 2018-08-03

### Commits

- Polymer3 Version [`b002c18`](https://github.com/drumr8403/xgd-table-of-contents/commit/b002c18fb9ab99b7b6a99ed2cc6a82eb55c3685c)
- added to collection [`e52c89f`](https://github.com/drumr8403/xgd-table-of-contents/commit/e52c89f221bed22004a3ca6f6938a8752a770404)
- fixed bower.json polymer resolution [`7ff660e`](https://github.com/drumr8403/xgd-table-of-contents/commit/7ff660e2779302da8d2470e3ea8426bdd3b70212)

## [1.0.2](https://github.com/drumr8403/xgd-table-of-contents/compare/1.0.1...1.0.2) - 2017-12-31

### Commits

- added paper-tooltip functionality [`02b4b2a`](https://github.com/drumr8403/xgd-table-of-contents/commit/02b4b2a999d19d8afb1f674d8c000cc3ed9c9272)

## [1.0.1](https://github.com/drumr8403/xgd-table-of-contents/compare/1.0.0...1.0.1) - 2017-12-31

### Commits

- added ability to have top-level pages [`9ac52c7`](https://github.com/drumr8403/xgd-table-of-contents/commit/9ac52c72dd89cd48f2e112f910d8537203b4fb74)

## 1.0.0 - 2017-12-31

### Commits

- updated documents for github [`c0eac6c`](https://github.com/drumr8403/xgd-table-of-contents/commit/c0eac6c03cacc2b4c896a77b127a91fa5a983762)
- Initial Commit to GitHub [`fd89631`](https://github.com/drumr8403/xgd-table-of-contents/commit/fd896310e429762a5aca92a8de6887d706ab2d94)
- Initial commit [`40d13b1`](https://github.com/drumr8403/xgd-table-of-contents/commit/40d13b1305874d5c0fba07241cc431e1c5a77117)


xgd-chapter.js
/**
@license
Copyright (c) 2018 XGDFalcon LLC. All rights reserved.
This code may only be used under the MIT style license found at https://github.com/XGDElements/xgd-table-of-contents/blob/master/LICENSE
Code distributed by Google as part of the polymer project is also
subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
*/
export class XGDChapter {
  constructor(source) {
    if (XGDChapter.exists(source)) {
      this.Title = source.Title;
      this.Chapters = source.Chapters;
      this.Link = source.Link;
      this.Element = source.Element;
    } else {
      this.Title = "NO TITLE";
      this.Chapters = [];
      this.Link = ""
    }
  }
  /** Get an identifier for this chapter object */
  GetIdentifier() {
    if (exists(Link)) {
      var start = Link.lastIndexOf('/') + 1;
      var end = Link.length;
      var link = Link.replace("#", "");
      return link.substring(start, end);
    } else {
      return Math.floor((1 + Math.random()) * 0x10000)
        .toString(16)
        .substring(1);
    }
  }
  /** Compare this object with a specific term */
  Compare(term) {
    if (XGDChapter.exists(this.Link)) {
      console.log(term);
      return this.Title.toUpperCase().indexOf(term.toUpperCase()) !== -1;
    } else {
      return true;
    }
  }
  /** 
   * Determines if the object actually exists
   *
   * @param Object item
   */
  static exists(item) {
    return (typeof (item) !== 'undefined' && item !== null && item !== "");
  }

  /**
   * Recursively searches for a link and opens all parent chapters.
   * @param {String} Link Link to search for.
   */
  static findLink(chapter, Link, chapters) {
    for(const item of chapter.Chapters) {
      if(item.Link) {
        if(item.Link === Link && (!chapter.Element || chapter.Element._hidden)) {
          for(const chapter2 of chapters) {
            chapter2.Element._toggleVisibility();
          }
        }
      } else {
        chapters.push(item);
        XGDChapter.findLink(item, Link, chapters);
      }
    }
  }
}

xgd-chapter-item.js
/**
@license
Copyright (c) 2018 XGDFalcon LLC. All rights reserved.
This code may only be used under the MIT style license found at https://github.com/XGDElements/xgd-table-of-contents/blob/master/LICENSE
Code distributed by Google as part of the polymer project is also
subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
*/
import { PolymerElement } from '@polymer/polymer/polymer-element.js';

import '@polymer/iron-icon/iron-icon.js';
import '@polymer/iron-icons/iron-icons.js';
import '@polymer/iron-flex-layout/iron-flex-layout.js';
import './xgd-page-item.js';
import { html } from '@polymer/polymer/lib/utils/html-tag.js';
import { afterNextRender } from '@polymer/polymer/lib/utils/render-status.js';
import {XGDChapter} from './xgd-chapter.js';
/**
 * `xgd-chapter-item`
 * Provides a table of contents for use in a document library or other such use.
 *
 * @customElement
 * @polymer
 * @demo demo/index.html
 */
class XGDChapterItem extends PolymerElement {
  static get template() {
    return html`
    <style>
      :host {
        display: block;
        @apply --layout-vertical;
        z-index: 0;
      }

      #content {
        @apply --layout-horizontal;
        @apply --layout-start-justified;
      }

      .title {
        @apply --layout-flex;
        font-size: var(--toc-font-size, 1em);
        height: auto;
        min-height: 1em;
      }

      a {
        font-size: var(--toc-font-size, 1em);
        text-decoration: none;
        height: auto;
        min-height: 1em;
      }

      #submenu {
        padding-left: 2ch;
        display: table-cell;
        @apply --layout-vertical;
      }

      iron-icon {
        margin-right: 1ch;
        --iron-icon-width: 36px;
        --iron-icon-height: 36px;
        --iron-icon-fill-color: var(--app-dark-primary-color, #000);
        --iron-icon-stroke-color: var(--app-dark-primary-color, #000);
      }

      #content.selected {
        background-color: var(--app-primary-color, #AA0);
      }

      #content.selected iron-icon {
        --iron-icon-fill-color: var(--app-lite-primary-color, #fff);
        --iron-icon-stroke-color: var(--app-lite-primary-color, #fff);
      }

      #content.selected .title {
        color: var(--light-text-color, #fff);
        font-weight: bold;
      }
    </style>

    <div id="content" on-click="_toggleVisibility">
      <iron-icon icon\$="{{_getIcon(_hidden)}}"></iron-icon>
      <span id="chaptername" class="title">[[chapter.Title]]</span>
    </div>
    <paper-tooltip animation-delay="200" for="chaptername" position="top">[[chapter.Title]]</paper-tooltip>
    <div id="submenu"></div>
`;
  }

  static get is() { return 'xgd-chapter-item'; }
  static get properties() {
    return {
      /** Determines if the submenu is visible */
      _hidden: {
        type: Boolean,
        value: true
      },
      /** Holds chapter information for this element */
      chapter: {
        type: XGDChapter
      },
      /** Holds the sections of this chapter */
      sections: {
        type: Array,
        value: []
      },
      /** Icon for closed folder */
      closed: {
        type: String,
        value: "icons:folder"
      },
      /** Icon for open folder */
      opened: {
        type: String,
        value: "icons:folder-open"
      },
      /** Icon for links */
      linkicon: {
        type: String,
        value: "icons:link"
      }
    };
  }

  /**
   * Handle a table of contents search
   *
   * @param {String} term
   */
  HandleSearch(term, level) {
    if (level) {
      // search children
      for (var x = 0; x < this.sections.length; x++) {
        var section = this.sections[x];
        if (section.tagName.toUpperCase() === "XGD-PAGE-ITEM") {
          section.HandleSearch(term);
        }
      }
    }
  }

  /**
   * Select this element if the title matches
   * @param {String} title 
   */
  SelectChapter(title, level) {
    if (level) {
      // search children
      for (var x = 0; x < this.sections.length; x++) {
        var section = this.sections[x];
        if (section.tagName.toUpperCase() === XGDChapterItem.is.toUpperCase()) {
          section.SelectChapter(title);
        }
      }

    }
    if (title === this.chapter.Title) {
      this.$.content.className = "selected";
      this._openParent();
    } else {
      this.$.content.className = "";
    }
  }

  /**
   * Select a particular link
   *
   * @param {String} link
   */
  SelectLink(link, level) {
    // search children
    if (level) {
      for (var x = 0; x < this.sections.length; x++) {
        var section = this.sections[x];
        section.SelectLink(link);
      }
    }
  }

  /**
   * Internal Use Only
   */
  _openParent() {
    this.ShowMenu(true);
    if (this.parentNode.parentNode.host) {
      this.parentNode.parentNode.host._openParent();
    }
  }

  /**
   * Opens or closes based on the toggle switch
   *
   * @param {Boolean} open
   */
  ShowMenu(open, level) {
    if (level) {
      for (var x = 0; x < this.sections.length; x++) {
        var section = this.sections[x];
        if (section.tagName.toUpperCase() === XGDChapterItem.is.toUpperCase()) {
          section.ShowMenu(open);
        }
      }
    }
    if (open) {
      this.$.submenu.style.display = 'table-cell';
    } else {
      this.$.submenu.style.display = 'none';
    }
    this.set('_hidden', !open);
  }

  /**
   * Get the proper icon if the submenu is visible/invisible
   *
   * @param {Boolean} hidden
   */
  _getIcon(hidden) {
    return hidden ? this.closed : this.opened;
  }

  /**
   * Handles button press to show/hide the submenu
   *
   * @param {Event} event
   */
  _toggleVisibility(event) {
    if (XGDChapter.exists(this.$.submenu)) {
      if (this._hidden) {
        this.$.submenu.style.display = 'table-cell';
        this.$.submenu.innerHTML = "";
        for (var x = 0; x < this.chapter.Chapters.length; x++) {
          if (XGDChapter.exists(this.chapter.Chapters[x].Link)) {
            var link = document.createElement("xgd-page-item");
            link.linkicon = this.linkicon;
            link.processChapter(this.chapter.Chapters[x], true);
            this.$.submenu.appendChild(link);
            this.push('sections', link);
          } else {
            var menu = document.createElement("xgd-chapter-item");
            menu.closed = this.closed;
            menu.opened = this.opened;
            menu.linkicon = this.linkicon;
            menu.processChapter(this.chapter.Chapters[x], true);
            this.$.submenu.appendChild(menu);
            this.push('sections', menu);
          }
        }
      } else {
        this.$.submenu.style.display = 'none';
      }
      this.set('_hidden', !this._hidden);
    }
  }

  /**
  * Process a chapter.
  * @param {XGDChapter} chapter
  */
  processChapter(chapter, toplevel) {
    chapter.Element = this;
    this.chapter = new XGDChapter(chapter);
  }
}

window.customElements.define(XGDChapterItem.is, XGDChapterItem);


xgd-table-of-contents.js
/**
@license
Copyright (c) 2018 XGDFalcon LLC. All rights reserved.
This code may only be used under the MIT style license found at https://github.com/XGDElements/xgd-table-of-contents/blob/master/LICENSE
Code distributed by Google as part of the polymer project is also
subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
*/
import { PolymerElement } from '@polymer/polymer/polymer-element.js';

import '@polymer/iron-icon/iron-icon.js';
import '@polymer/iron-icons/iron-icons.js';
import '@polymer/paper-input/paper-input.js';
import '@polymer/paper-toggle-button/paper-toggle-button.js';
import '@polymer/iron-flex-layout/iron-flex-layout.js';
import './xgd-chapter-item.js';
import './xgd-page-item.js';
import { html } from '@polymer/polymer/lib/utils/html-tag.js';
import { afterNextRender } from '@polymer/polymer/lib/utils/render-status.js';
import {XGDChapter} from './xgd-chapter.js';
/**
 * `xgd-table-of-contents`
 * Provides a table of contents for use in a document library or other such use.
 *
 * @customElement
 * @polymer
 * @demo demo/index.html
 */
class XGDTableOfContents extends PolymerElement {
  static get template() {
    return html`
    <style>
      :host {
        display: block;
      }

      #toc {
        @apply --layout-vertical;
        @apply --layout-start-justified;
      }
    </style>
    <div class="content">
      <div class="search">
        <paper-input label="Search Contents" value="{{_searchTerm}}">
          <iron-icon icon="icons:search" slot="prefix"></iron-icon>
          <iron-icon icon="icons:clear" slot="suffix" on-click="_clearSearch"></iron-icon>
        </paper-input>
      </div>
      <div class="container">
        <paper-toggle-button checked="" on-click="_handleToggle">[[toggleState]]</paper-toggle-button>
        <div id="toc">
        </div>
      </div>
    </div>
`;
  }

  static get is() { return 'xgd-table-of-contents'; }
  static get properties() {
    return {
      _searchTerm: {
        type: String,
        observer: "_searchTable"
      },
      _tableOfContents: {
        type: Array,
        value: []
      },
      /** Icon for closed folder */
      closed: {
        type: String,
        value: "icons:folder"
      },
      /** Icon for open folder */
      opened: {
        type: String,
        value: "icons:folder-open"
      },
      /** Icon for links */
      linkicon: {
        type: String,
        value: "icons:link"
      }
    };
  }

  /**
   * Recursively searches for a link and opens all parent chapters.
   * @param {String} Link Link to search for.
   */
  findLink(Link) {
    for(const item of this._tableOfContents) {
      if(!item.chapter.Link) {
        XGDChapter.findLink(item.chapter, Link, [item.chapter]);
      }
    }
  }

  /**
   * Fired when a user selects a page.
   * @event xgd-page-selection
   * @param {Event} event 
   *
   * The structure of the event is as follows:
   * @param {Object} detail {
   *      @param {String} Link Link to the page.
   * }
   */

  constructor() {
    super();
    afterNextRender(this, function () {
      window.addEventListener("xgd-page-selection", function (event) {
        this.findLink(event.detail.Link);
        this.SelectPage(event.detail.Link);
      }.bind(this));
    });
  }
  /**
  * Clear the search box.
  * @param {Event} event
  */
  _clearSearch(event) {
    this.set('_searchTerm', "");
  }

  /**
  * Search the table of contents.
  * @param {String} term
  */
  _searchTable(term) {
    for (var x = 0; x < this._tableOfContents.length; x++) {
      if (this._tableOfContents[x].HandleSearch) {
        this._tableOfContents[x].HandleSearch(term, true);
      }
    }
  }

  /**
  * Select a particular page.
  * @param {String} pagelink
  */
  SelectPage(pagelink) {
    console.log('Inside SelectPage');
    for (var x = 0; x < this._tableOfContents.length; x++) {
      if (this._tableOfContents[x].SelectLink) {
        console.log('Inside if(...SelectLink');
        this._tableOfContents[x].SelectLink(pagelink, true);
      }
    }
  }

  /**
  * Select a particular chapter.
  * @param {String} title
  */
  SelectChapter(title) {
    for (var x = 0; x < this._tableOfContents.length; x++) {
      if (this._tableOfContents[x].SelectChapter) {
        this._tableOfContents[x].SelectChapter(title, true);
      }
    }
  }

  /**
  * Handles the even from the toggle button.
  * @param {Event} event
  */
  _handleToggle(event) {
    var open = event.target.checked;
    this._toggleMenu(open);
  }

  /**
  * Toggle expand/collapse of the table of contents.
  * @param {Event} event
  */
  _toggleMenu(open) {
    for (var x = 0; x < this._tableOfContents.length; x++) {
      if (this._tableOfContents[x].ShowMenu) {
        this._tableOfContents[x].ShowMenu(open, true);
      }
    }
  }
  /**
  * Set the Table Of Contents Data.
  * @param {Array}(XGDChapter) toc
  */
  LoadTableOfContents(toc) {
    if (toc !== null && typeof(toc) !== 'undefined' && typeof(toc.Chapters) !== 'undefined') {
      this.$.toc.innerHTML = "";
      for (var i = 0; i < toc.Chapters.length; i++) {
        if (XGDChapter.exists(toc.Chapters[i].Link)) {
          var link = document.createElement("xgd-page-item");
          link.linkicon = this.linkicon;
          link.processChapter(toc.Chapters[i]);
          this.$.toc.append(link);
          this.push("_tableOfContents", link);
          if(i === 0 && link.SelectLink) {
            link.SelectLink(link.chapter.Link);
          }
        } else {
          var chapter = document.createElement("xgd-chapter-item");
          chapter.closed = this.closed;
          chapter.opened = this.opened;
          chapter.linkicon = this.linkicon;
          chapter.processChapter(toc.Chapters[i]);
          this.$.toc.append(chapter);
          this.push("_tableOfContents", chapter);
        }
      }
    }
  }
}

window.customElements.define(XGDTableOfContents.is, XGDTableOfContents);


xgd-page-item.js
/**
@license
Copyright (c) 2018 XGDFalcon LLC. All rights reserved.
This code may only be used under the MIT style license found at https://github.com/XGDElements/xgd-table-of-contents/blob/master/LICENSE
Code distributed by Google as part of the polymer project is also
subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
*/
import { PolymerElement } from '@polymer/polymer/polymer-element.js';

import '@polymer/iron-icon/iron-icon.js';
import '@polymer/iron-icons/iron-icons.js';
import '@polymer/paper-tooltip/paper-tooltip.js';
import { html } from '@polymer/polymer/lib/utils/html-tag.js';
import { afterNextRender } from '@polymer/polymer/lib/utils/render-status.js';
import {XGDChapter} from './xgd-chapter.js';

/**
 * `xgd-page-item`
 * Provides a table of contents for use in a document library or other such use.
 *
 * @customElement
 * @polymer
 * @demo demo/index.html
 */
class XGDPageItem extends PolymerElement {
  static get template() {
    return html`
    <style>
      :host {
        display: block;
        z-index: 0;
      }

      #content {
        @apply --layout-horizontal;
        @apply --layout-start-justified;
        cursor: pointer;
        background-color: transparent;
        color: var(--primary-text-color, #000);
        font-style: normal;
      }

      .title {
        @apply --layout-flex;
        font-size: var(--toc-font-size, 1em);
        text-decoration: none;
        height: auto;
        min-height: 1em;
      }

      iron-icon {
        margin-right: 1ch;
        --iron-icon-width: 36px;
        --iron-icon-height: 36px;
        --iron-icon-fill-color: var(--app-dark-primary-color, #000);
        --iron-icon-stroke-color: var(--app-dark-primary-color, #000);
      }

      #content.selected {
        background-color: var(--app-primary-color, #AA0);
      }

      #content.selected iron-icon {
        --iron-icon-fill-color: var(--app-lite-primary-color, #fff);
        --iron-icon-stroke-color: var(--app-lite-primary-color, #fff);
      }

      #content.selected .title {
        color: var(--light-text-color, #fff);
        font-weight: bold;
      }
    </style>

    <div id="content">
      <iron-icon id="icon" icon\$="[[linkicon]]"></iron-icon>
      <span id="pagename" class="title">[[chapter.Title]]</span>
    </div>
    <paper-tooltip animation-delay="200" for="pagename" position="top">[[chapter.Title]]</paper-tooltip>
`;
  }

  static get is() { return 'xgd-page-item'; }
  static get properties() {
    return {
      /** Determines if the submenu is visible */
      _hidden: {
        type: Boolean,
        value: true
      },
      /** Holds chapter information for this element */
      chapter: {
        type: XGDChapter
      },
      /** Icon for links */
      linkicon: {
        type: String,
        value: "icons:link"
      }
    };
  }

  constructor() {
    super();
    afterNextRender(this, function () {
      this.$.content.addEventListener('click', this._handleClick.bind(this));
    });
  }

  /**
   * Select this element if the link matches
   * @param {String} link 
   */
  SelectLink(link) {
    if (link === this.chapter.Link) {
      this.$.content.className = "selected";
      if (this.parentNode && this.parentNode.parentNode.host) {
        this.parentNode.parentNode.host._openParent();
      }
      this.scrollIntoView();
    } else {
      this.$.content.className = "";
    }
  }
  /**
   * Handle clicking this item
   *
   * @param {String} term
   */
  _handleClick(event) {
    var chapter = new XGDChapter(this.chapter);
    window.dispatchEvent(new CustomEvent("xgd-page-selection", { detail: chapter, bubbles: true }));
  }

  /**
   * Handle a table of contents search
   *
   * @param {String} term
   */
  HandleSearch(term) {
    if (!this.chapter.Compare(term)) {
      this.style.display = "none";
    } else {
      this.style.display = "block";
    }
  }

  /**
  * Process a chapter.
  * @param {XGDChapter} chapter
  */
  processChapter(chapter) {
    this.chapter = new XGDChapter(chapter);
  }
}

window.customElements.define('xgd-page-item', XGDPageItem);
